pkg/ccl/backupccl/file_sst_sink_test.go

```diff
@@ -135,3 +141,819 @@ func TestFileSSTSinkExtendOneFile(t *testing.T) {
+		{
+			// Write the first exported span after the flush.
+			newExportedSpanBuilder("i", "k", true).withKVs([]kvAndTS{{key: "i", timestamp: 10}, {key: "j", timestamp: 10}}).build(),
+			sinkStats{hlc.Timestamp{}, 1, 5, 1, 0, 1, 1}},
+		{
+			// Write another exported span that causes an out of order flush.
+			newExportedSpanBuilder("j", "l", true).withKVs([]kvAndTS{{key: "j", timestamp: 10}, {key: "k", timestamp: 10}}).build(),
+			sinkStats{hlc.Timestamp{}, 1, 6, 2, 1, 1, 1}},
+	}
+
+	for _, input := range inputs {
+		require.NoError(t, sink.write(ctx, input.input))
+
+		actualStats := sinkStats{
+			flushedRevStart: sink.flushedRevStart,
+			completedSpans:  sink.completedSpans,
+			files:           sink.stats.files,
+			flushes:         sink.stats.flushes,
+			oooFlushes:      sink.stats.oooFlushes,
+			sizeFlushes:     sink.stats.sizeFlushes,
+			spanGrows:       sink.stats.spanGrows,
+		}
+
+		require.Equal(t, input.expectedStats, actualStats, "stats after write for span %v", input.input.metadata.Span)
+	}
+
+	require.NoError(t, sink.flush(ctx))
+}
+
+func TestFileSSTSinkCopyPointKeys(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+	settings := cluster.MakeTestingClusterSettings()
+
+	type testInput struct {
+		input     []kvAndTS
+		expectErr string
+	}
+
+	type testCase struct {
+		name   string
+		inputs []testInput
+	}
+
+	for _, tt := range []testCase{
+		{
+			name: "single-sst",
+			inputs: []testInput{
+				{
+					input: []kvAndTS{
+						{key: "a", value: []byte("1"), timestamp: 10},
+						{key: "a", value: []byte("2"), timestamp: 9},
+						{key: "b", value: []byte("2"), timestamp: 10},
+					},
+				},
+			},
+		},
+		{
+			name: "multiple-ssts",
+			inputs: []testInput{
+				{
+					input: []kvAndTS{
+						{key: "a", value: []byte("1"), timestamp: 10},
+						{key: "a", value: []byte("2"), timestamp: 9},
+						{key: "b", value: []byte("2"), timestamp: 10},
+					},
+				},
+				{
+					input: []kvAndTS{
+						{key: "c", value: []byte("3"), timestamp: 10},
+						{key: "d", value: []byte("4"), timestamp: 9},
+						{key: "e", value: []byte("5"), timestamp: 8},
+					},
+				},
+				{
+					input: []kvAndTS{
+						{key: "e", value: []byte("3"), timestamp: 6},
+						{key: "f", value: []byte("4"), timestamp: 10},
+						{key: "g", value: []byte("5"), timestamp: 10},
+					},
+				},
+			},
+		},
+		{
+			name: "out-of-order-key",
+			inputs: []testInput{
+				{
+					input: []kvAndTS{
+						{key: "a", value: []byte("1"), timestamp: 10},
+						{key: "c", value: []byte("2"), timestamp: 10},
+					},
+				},
+				{
+					input: []kvAndTS{
+						{key: "b", value: []byte("3"), timestamp: 10},
+						{key: "d", value: []byte("4"), timestamp: 10},
+					},
+					expectErr: "keys must be added in strictly increasing order",
+				},
+			},
+		},
+		{
+			name: "out-of-order-timestamp",
+			inputs: []testInput{
+				{
+					input: []kvAndTS{
```

Comment:
looks like the timestamps are all the same here

Reply:
oh interesting. I don't even know if it's possible for a mvcc key to have multiple values at a given timestamp? Might be worth checking with kv peeps, and if it's not possible, just change that second kv on `b` to a timestamp of 9.

Reply:
I change the second value to something different (11 instead of 9)